Updates and New Collectors

kamaev opened the pull request.

kamaev committed:
update smartcollector

kamaev committed:
added more cases for tests
revert

beevee commented on docs/collectors/SmartCollector.md:
```diff
@@ -15,35 +15,37 @@ Collect data from S.M.A.R.T.'s attribute reporting.
 bin | smartctl | The path to the smartctl binary | str
 byte_unit | byte | Default numeric output(s) | str
-devices | ^disk[0-9]$|^sd[a-z]$|^hd[a-z]$ | device regex to collect stats on | str
+devices | ['sda'] | Devices to collect stats on | list
```
Тут раньше были регулярки, а теперь только строки можно? Регулярки же лучше? :)

kamaev replied:
Вернул обработку регулярки

beevee commented on docs/collectors/SmartCollector.md:
```diff
@@ -15,35 +15,37 @@ Collect data from S.M.A.R.T.'s attribute reporting.
 byte_unit | byte | Default numeric output(s) | str
-devices | ^disk[0-9]$|^sd[a-z]$|^hd[a-z]$ | device regex to collect stats on | str
+devices | ['sda'] | Devices to collect stats on | list
+force_prefails | False | Fetch prefails anyway | bool
```
Хорошо бы тут получше объяснение написать, что происходит в случае True и False в этом параметре. Например, что-то вроде:

If true, fetch all prefail and whitelisted old-age attributes. Otherwise, fetch only whitelisted prefail and old-age attributes.

kamaev replied:
Переписал описание, только без упоминания whitelists, чтобы не возникло путаницы с нативными whitelist'ами (регулярками). Если что, перепишу

beevee commented on docs/collectors/SmartCollector.md:
```diff
@@ -15,35 +15,37 @@ Collect data from S.M.A.R.T.'s attribute reporting.
-servers.hostname.smart.sda.Start_Stop_Count 8
-servers.hostname.smart.sda.Temperature_Celsius 28
-servers.hostname.smart.sda.UDMA_CRC_Error_Count 0
+path_prefix.hostname.sda.pre-fail.value.raw_read_error_rate 100
```
Почему именно `value.raw_read_error_rate`, а не `raw_read_error_rate.value`? Второе кажется более стандартным, например statsd кладет `parametername.count` или `parametername.p95`, а не наоборот.

kamaev replied:
Сменил порядок с attribute.valtype на valtype.attribute

beevee commented on src/collectors/smart/smart.py:
```diff
@@ -114,3 +102,55 @@ def find_attr_start_line(self, lines, min_line=4, max_line=9):
+        Fetching S.M.A.R.T. data.
+        """
+        if self.config['force_prefails']:
+            critical = 'pre-fail'
```
Неудачное имя переменной, лучше что-нибудь вроде `force_fetch_priority`.

kamaev replied:
Переименовал на force_fetch_priority. Перенёс все требования к аттрибутам  в 'танатос' класса (__new__)

beevee commented on src/collectors/smart/smart.py:
```diff
@@ -9,23 +9,63 @@
+    def __init__(self,
+                 instance):
+
+        self.instance = instance
```
Неудачное имя переменной. Лучше что-то вроде `raw_attribute` или `attribute_data`.

kamaev replied:
Переименовал на attribute_data

kamaev committed:
changed broken link to smartmontools wiki

kamaev committed:
added new collector DirStatsCollector

kamaev committed:
optimized dirstats collectors

- added total files counter
- changed os.getlogin to getpass.getuser to prevent oserrors for several
cases

kamaev committed:
dirstats (code, docs) + smart (docs) update

dirstats: added subdirs counters
dirstats, smart: updated docs

kamaev committed:
improved dirstats collector

kamaev commented:
Ready for review

kamaev committed:
added space to logstr

beevee commented on docs/collectors/SmartCollector.md:
```diff
@@ -8,42 +8,44 @@ Collect data from S.M.A.R.T.'s attribute reporting.
-servers.hostname.smart.sda.Start_Stop_Count 8
-servers.hostname.smart.sda.Temperature_Celsius 28
-servers.hostname.smart.sda.UDMA_CRC_Error_Count 0
+servers.hostname.sda.pre-fail.raw_read_error_rate.value 100
```
А чего smart убрал? Вроде другие коллекторы вставляют название коллектора в имя метрики (например, ntpd, mongodb и так далее).

kamaev replied:
Upd: Исправил пути для коллекторов

beevee commented on src/collectors/dirstats/dirstats.py:
```diff
@@ -0,0 +1,139 @@
+
+import diamond.collector
+
+BLOCK = 512
```
А размер блока всегда 512?

kamaev replied:
Это значение занесено в константу по двум причинам:

>st_blocks - number of 512-byte blocks allocated for file
(https://docs.python.org/2.7/library/os.html)

>here is no correlation between values of the st_blocks and st_blksize, and the f_bsize (from <sys/statvfs.h>) structure members.
>>Under Linux, 512 turns out to be the right multiplier (and not st_blksize)
(https://bugs.python.org/issue10016)

Т.е. нет какого то более гибкого выражения вроде A*B чтобы посчитать размер файла в абсолютно всех случаях. Можно конечно пользоваться только os.stat(fullpath).st_size, но тогда не будет возможности считать размер sparse-файлов, вроде тех-же графитовских *wsp, размера виртуальных жестких дисков и пр.

kamaev replied:
В доке к python 3.6 уже просто вот так пишут:
>st_blocks
Number of 512-byte blocks allocated for file. This may be smaller than st_size/512 when the file has holes.

beevee replied:
Понял, ок.

kamaev committed:
added collectors paths
